bpf/lib/egress_policies.h

```diff
@@ -579,25 +582,87 @@ srv6_store_meta_sid(struct __ctx_buff *ctx, const union v6addr *sid)
 	ctx_store_meta(ctx, CB_SRV6_SID_4, sid->p4);
 }
 
+#ifdef ENABLE_IPV6
+/* SRv6 encapsulation occurs at the native-dev currently.
+ * Its possible that after encapsulation a fib entry exists which would actually
+ * route the IPv6 destination somewhere else.
+ *
+ * Therefore, this function performs an additional fib lookup on the encap'd
+ * packet to ensure we transmit it via the correct link and with the correct
+ * l2 addresses.
+ */
+static __always_inline int
+srv6_refib(struct __ctx_buff *ctx, int *ext_err)
+{
+	struct bpf_fib_lookup_padded params = {0};
+	__u32 old_oif = ctx_get_ifindex(ctx);
+	void *data, *data_end;
+	struct ipv6hdr *ip6;
+
+	if (!revalidate_data(ctx, &data, &data_end, &ip6))
+		return DROP_INVALID;
+
+	*ext_err = (__s8)fib_lookup_v6(ctx, &params, ip6, BPF_FIB_LOOKUP_OUTPUT);
+
+	switch (*ext_err) {
+	case BPF_FIB_LKUP_RET_SUCCESS:
+		/* We found an oif and ARP was successful.
+		 * We may need to redirect to the appropriate oif, if not
+		 * rewrite the layer 2 and continue processing.
+		 */
+		if (old_oif != params.l.ifindex)
+			return fib_do_redirect(ctx, true, &params,
+					      (__s8 *)ext_err, (int *)&old_oif);
+
+		if (eth_store_daddr(ctx, params.l.dmac, 0) < 0)
+			return DROP_WRITE_ERROR;
+
```

Comment:
This is needed as you pushed the SRv6 encap headers onto the packet, but nothing has actually filled the outer L2 header yet. Right? Where would the source MAC come from in this case?

(I don't have a full picture of the SRv6 path, so this might be obvious ...).

Reply:
Kinda, 

Traffic leaves the pod right, not tunneled. 

It hits the host net stack, where a FIB lookup is done on the **outer packet** and the l2 is decided then along with output interface, its then pushed to the output interface. 

On the output interface we have SRv6 code running. 
The encap actually just pushes the **existing**  l2 done for the outer packet forward, and does the IPIP encap between the the original l2 and the original l3 headers. 

So, we were just keeping the original l2 header around. As an additional point, this also means we were always sending out the same native dev too, even if the encap'd packet should route on a different link.

Reply:
ah! And this is the case where we stay on the same egress interface, so the initial smac is still valid. All good then :).

Reply:
Yup! and if we redirect, we re-write both :)